### PR TITLE
Plugin timeout

### PR DESCRIPTION
**Pull request by mcollina**
This would help solving one of the major pain points of using Fastify, developers not calling their callback and having trouble to debug where the the loading process have stopped.

### CODE REVIEW

**Comment by mcollina** [comment]
cc @davidmarkclements

**Comment by davidmarkclements** [comment]
yes - I can attest this caused some issues

does the timeout have to be zero by default? 

I suppose so, since it depends on plugins for acceptable timeouts.

But what if the default time out was 500ms, and plugins could define their own timeout to override?

File: boot.js [hunk 1]
```diff
@@ -130,12 +131,12 @@ function Boot (server, opts, done) {
   }
 
   this._doStart = null
-  const main = new Plugin(this, (s, opts, done) => {
+  const main = new Plugin(this, function root (s, opts, done) {
```
Comment (delvedor) [inline boot.js]: Why did you remove the arrow function?

File: plugin.js [hunk 2]
```diff
@@ -42,6 +63,15 @@ Plugin.prototype.exec = function (server, cb) {
 
   debug('exec', name)
 
+  var timer
+
+  if (this.parent._timeout > 0) {
+    timer = setTimeout(function () {
+      timer = null
```
Comment (delvedor) [inline plugin.js]: Can you add a debug log here?

File: boot.js (hunk 1 above)
Comment (mcollina) [reply boot.js]: So I get `root` as the name in the debug.

File: plugin.js [hunk 3]
```diff
@@ -42,6 +63,15 @@ Plugin.prototype.exec = function (server, cb) {
...
+      done(new Error('plugin did not start in time: ' + name))
```
Comment (delvedor) [inline plugin.js]: I would rephrase this as `plugin '${name}' timed out.`

Comment (mcollina) [reply plugin.js]: I didn't do that because name can be quite a long string (the first two lines of the function in the bad case). Having those in the middle would make the error unreadable.

**Comment by davidmarkclements** [comment]
Looking awesome 👏

**Comment by mcollina** [comment]
Done.
